### PR TITLE
feat: Add Firestore backup/schedule and restore support

### PR DESCRIPTION
**Pull request by rwhogg**
### Description

Creates new commands for Firestore disaster recovery support:

* `firestore:backups:schedules:list`
* `firestore:backups:schedules:create`
* `firestore:backups:schedules:update`
* `firestore:backups:schedules:delete`
* `firestore:backups:get`
* `firestore:backups:list`
* `firestore:backups:delete`
* `firestore:databases:restore`

### Scenarios Tested

`firestore:backups:schedules:list`
* [x] Successful list (no schedules)
* [x] Successful list (daily schedule only)
* [x] Successful list (weekly schedule only)
* [x] Successful list (both)
* [x] Database flag defaults to "(default)"
* [x] Unsuccesful list (database does not exist)

`firestore:backups:schedules:create`
* [x] Successful creation (daily alone)
* [x] Successful creation (weekly alone)
* [x] Successful creation (daily with preexisting weekly)
* [x] Successful creation (weekly with preexisting daily)
* [x] Unsuccessful creation (two dailies)
* [x] Unsuccessful creation (two weeklies)
* [x] Unsuccessful creation (database does not exist)
* [x] Database flag defaults to "(default)"
* [x] Client error (no retention specified)
* [x] Client error (no recurrence specified)
* [x] Client error (bad retention syntax)
* [x] Client error (recurrence other than DAILY or WEEKLY)
* [x] Client error (weekly retention but no day of week)
* [x] Client error (day of week specified for daily retention)
* [x] Unsuccessful creation (daily retention, >7 days)
* [x] Unsuccessful creation (weekly retention, >14 weeks)

`firestore:backups:schedules:update`
* [x] Successful update (daily)
* [x] Successful update (weekly)
* [x] Unsuccessful update (schedule does not exist)
* [x] Client error (try to update recurrence)
* [x] Client error (no retention specified)
* [x] Client error (bad retention syntax)
* [x] Unsuccessful update (daily retention, >7 days)
* [x] Unsuccessful update (weekly retention, >14 weeks)

`firestore:backups:schedules:delete`
* [x] Successful delete (daily)
* [x] Successful delete (weekly)
* [x] Unsuccessful delete (schedule does not exist)
* [x] Delete aborted (confirmation prompt rejected)
* [x] Delete allowed (confirmation prompt allowed)
* [x] Confirmation prompt skipped (--force)

`firestore:backups:get`
* [x] Successful get
* [x] Unsuccessful get (backup does not exist)
* [x] Client error (backup not specified)

`firestore:backups:list`
* [x] Successful list (no backups)
* [x] Successful list (> 0 backups)
* [x] Location flag defaults to "-" ("all locations")
* [x] Unsuccessful list (location does not exist)

`firestore:backups:delete`
* [x] Successful delete
* [x] Unsuccessful delete (backup does not exist)
* [x] Client error (backup not specified)
* [x] Delete aborted (confirmation prompt rejected)
* [x] Delete allowed (confirmation prompt allowed)
* [x] Confirmation prompt skipped (--force)

`firestore:databases:restore`
* [x] Successful restore initiation
* [x] Unsuccessful restore initiation (backup does not exist)
* [x] Unsuccessful restore initiation (database name is already taken)
* [x] Client error (backup not specified)
* [x] Client error (database name not specified)

### Sample Commands

* `firebase firestore:backups:get projects/myproj/locations/nam5/backups/abcdefg`
* `firebase firestore:backups:list –location=nam5 –project=myproj`
* `firebase firestore:backups:delete projects/myproj/locations/nam5/backup/abcdefg`
* `firebase firestore:backups:schedules:list --database '(default)'`
* `firebase firestore:backups:schedules:create --database '(default)' --recurrence DAILY --retention 3d`
* `firebase firestore:backups:schedules:create --database '(default)' --recurrence WEEKLY --day-of-week TUESDAY --retention 7d`
* `firebase firestore:backups:schedules:update projects/myproj/databases/(default)/backupSchedules/hijklmnop --retention 4d`
* `firebase firestore:backups:schedules:delete projects/myproj/databases/(default)/backupSchedules/hijklmnop`
* `firebase firestore:databases:restore --database restore-target --backup projects/myproject/locations/nam5/backups/abcdefg`

### CODE REVIEW

File: src/firestore/api-sort.ts [hunk 1]
```diff
@@ -81,6 +82,23 @@ export function compareLocation(a: API.Location, b: API.Location): number {
...
+  // Name should always be unique and present
+  return a.name! > b.name! ? 1 : -1;
+}
+
+/**
+ * Compare two BackupSchedule API entries for sorting.
+ */
+export function compareApiBackupSchedule(a: BackupSchedule, b: BackupSchedule): number {
+  // Name should always be unique and present
+  // FIXME: IS THIS HOW WE WANT TO SORT???
```
Comment (rwhogg) [inline src/firestore/api-sort.ts:98]: FIXME: check with the rest of the team what the sorting should be

Comment (rwhogg) [reply src/firestore/api-sort.ts:98]: (for both backups and schedules)

File: src/firestore/api.ts [hunk 2]
```diff
@@ -329,6 +330,122 @@ export class FirestoreApi {
...
+   */
+  prettyPrintBackup(backup: Backup) {
+    const Table = require("cli-table");
+    const table = new Table({
+      head: ["Field", "Value"],
+      colWidths: [25, Math.max(50, 5 + backup.name!.length, 5 + backup.database!.length)],
+    });
+
+    table.push(
+      // FIXME: is everything supposed to be yellow??? same question about backup schedules
```
Comment (rwhogg) [inline src/firestore/api.ts:437]: FIXME: check what the database commands do

Comment (rwhogg) [reply src/firestore/api.ts]: It's consistent

**Comment by rwhogg** [comment]
Requesting review from @zixuanliucanada

**Comment by ZixuanLiuCanada** [comment]
LGTM with some comments. Nice job! I learned something new today.

A qq, can you please point me where these tests are? I saw you have a comprehensive test coverage. But I cannot find the code.

**Comment by rwhogg** [comment]
This was the result of manual testing; Firestore support here in general would benefit from automation.

**Comment by rwhogg** [comment]
Closing in favor of #6778 6778

**Pull request closed without merging**